File: budget/frontend/src/lib/splits.ts
/**
 * Auflösen der Aufteilungsvorlagen im Frontend.
 *
 * Spiegelt `backend/app/services/splits.py`. Der Server bleibt die Instanz, die
 * validiert und speichert — diese Kopie existiert nur, damit die Vorschau im
 * Erfassungsformular ohne Netzwerk-Roundtrip sofort dasselbe Ergebnis zeigt.
 */

import type { Member, SplitLine, SplitTemplate } from "@/api/types";
import { allocate } from "./money";

export interface SplitResult {
  lines: SplitLine[];
  /** Was noch fehlt, damit die Summe stimmt (nur bei MANUAL relevant). */
  remainderMinor: number;
  valid: boolean;
  error?: string;
}

export function resolveSplit(
  template: SplitTemplate,
  totalMinor: number,
  activeMembers: Member[],
  options: { singleMemberId?: number | null; manual?: SplitLine[] } = {},
): SplitResult {
  const ordered = [...activeMembers].sort((a, b) => a.sort_order - b.sort_order || a.id - b.id);

  if (totalMinor === 0) {
    return { lines: [], remainderMinor: 0, valid: false, error: "Der Betrag darf nicht 0 sein." };
  }

  if (template === "SINGLE") {
    const memberId = options.singleMemberId ?? ordered[0]?.id;
    if (!memberId) {
      return { lines: [], remainderMinor: totalMinor, valid: false, error: "Keine Person gewählt." };
    }
    return { lines: [{ member_id: memberId, amount_minor: totalMinor }], remainderMinor: 0, valid: true };
  }

  if (template === "EQUAL" || template === "KEY") {
    if (ordered.length === 0) {
      return { lines: [], remainderMinor: totalMinor, valid: false, error: "Keine aktiven Personen." };
    }
    const weights = template === "EQUAL" ? ordered.map(() => 1) : ordered.map((m) => m.share_weight);
    if (weights.reduce((a, b) => a + b, 0) <= 0) {
      return { lines: [], remainderMinor: totalMinor, valid: false, error: "Verteilschlüssel ist leer." };
    }
    const amounts = allocate(totalMinor, weights);
    const lines = ordered
      .map((member, index) => ({ member_id: member.id, amount_minor: amounts[index] }))
      .filter((line) => line.amount_minor !== 0);
    return { lines, remainderMinor: 0, valid: lines.length > 0 };
  }

  const manual = (options.manual ?? []).filter((line) => line.amount_minor !== 0);
  const sum = manual.reduce((total, line) => total + line.amount_minor, 0);
  const remainder = totalMinor - sum;
  const mixedSigns = manual.some((line) => line.amount_minor > 0) && manual.some((line) => line.amount_minor < 0);

  if (mixedSigns) {
    return {
      lines: manual,
      remainderMinor: remainder,
      valid: false,
      error: "Alle Anteile müssen dasselbe Vorzeichen haben.",
    };
  }
  return {
    lines: manual,
    remainderMinor: remainder,
    valid: remainder === 0 && manual.length > 0,
  };
}

/** Die Aufteilung einer bestehenden Buchung als Vorlage erkennen (für die Bearbeitung). */
export function detectTemplate(splits: SplitLine[], activeMembers: Member[]): SplitTemplate {
  if (splits.length === 1) return "SINGLE";
  const total = splits.reduce((sum, line) => sum + line.amount_minor, 0);
  for (const template of ["EQUAL", "KEY"] as const) {
    const candidate = resolveSplit(template, total, activeMembers);
    if (sameLines(candidate.lines, splits)) return template;
  }
  return "MANUAL";
}

function sameLines(a: SplitLine[], b: SplitLine[]): boolean {
  if (a.length !== b.length) return false;
  const map = new Map(b.map((line) => [line.member_id, line.amount_minor]));
  return a.every((line) => map.get(line.member_id) === line.amount_minor);
}

File: budget/backend/app/services/splits.py
"""Aufteilungsvorlagen.

Vorlagen sind ausschliesslich ein Eingabe-Helfer: aufgeloest wird immer sofort in
konkrete Betraege. Gespeichert werden nur die Splits, damit eine spaetere Aenderung
am Verteilschluessel alte Buchungen nicht rueckwirkend verfaelscht.
"""

from __future__ import annotations

from dataclasses import dataclass

from app.enums import SplitTemplate
from app.models import Member
from app.services.money import allocate


class SplitError(ValueError):
    """Fachlicher Fehler bei der Aufteilung (fuehrt im API-Layer zu HTTP 422)."""


@dataclass(frozen=True, slots=True)
class SplitLine:
    member_id: int
    amount_minor: int


def _ordered(members: list[Member]) -> list[Member]:
    return sorted(members, key=lambda m: (m.sort_order, m.id))


def build_splits(
    template: SplitTemplate,
    total_minor: int,
    members: list[Member],
    *,
    single_member_id: int | None = None,
    manual: list[SplitLine] | None = None,
) -> list[SplitLine]:
    """Loest eine Vorlage in konkrete Splits auf.

    ``members`` sind die **aktiven** Personen des Haushalts; deaktivierte Personen
    bekommen keine neuen Splits mehr, behalten ihre alten aber.
    """
    if total_minor == 0:
        raise SplitError("Der Betrag darf nicht 0 sein.")

    if template is SplitTemplate.SINGLE:
        if single_member_id is None:
            raise SplitError("Fuer die Vorlage 'Eine Person' fehlt die Person.")
        if single_member_id not in {m.id for m in members}:
            raise SplitError("Die gewaehlte Person gehoert nicht zu den aktiven Personen.")
        return [SplitLine(single_member_id, total_minor)]

    if template is SplitTemplate.EQUAL:
        active = _ordered(members)
        if not active:
            raise SplitError("Der Haushalt hat keine aktiven Personen.")
        amounts = allocate(total_minor, [1] * len(active))
        return [SplitLine(m.id, a) for m, a in zip(active, amounts, strict=True)]

    if template is SplitTemplate.KEY:
        active = _ordered(members)
        if not active:
            raise SplitError("Der Haushalt hat keine aktiven Personen.")
        weights = [m.share_weight for m in active]
        if sum(weights) <= 0:
            raise SplitError("Der Verteilschluessel des Haushalts ist leer.")
        amounts = allocate(total_minor, weights)
        return [SplitLine(m.id, a) for m, a in zip(active, amounts, strict=True) if a != 0]

    if template is SplitTemplate.MANUAL:
        if not manual:
            raise SplitError("Fuer die manuelle Aufteilung fehlen die Betraege.")
        lines = [line for line in manual if line.amount_minor != 0]
        if not lines:
            raise SplitError("Mindestens ein Split muss einen Betrag ungleich 0 haben.")
        validate(total_minor, lines)
        return lines

    raise SplitError(f"Unbekannte Aufteilungsvorlage: {template}")


def validate(total_minor: int, lines: list[SplitLine]) -> None:
    """Prueft die Invarianten, die auch die DB-Trigger erzwingen -- nur mit besserer Meldung."""
    if not lines:
        raise SplitError("Eine Buchung braucht mindestens einen Split.")
    if any(line.amount_minor == 0 for line in lines):
        raise SplitError("Ein Split mit Betrag 0 ist nicht zulaessig.")
    signs = {line.amount_minor > 0 for line in lines}
    if len(signs) > 1:
        raise SplitError("Alle Splits einer Buchung muessen dasselbe Vorzeichen haben.")
    seen: set[int] = set()
    for line in lines:
        if line.member_id in seen:
            raise SplitError("Pro Person ist hoechstens ein Split je Buchung erlaubt.")
        seen.add(line.member_id)
    actual = sum(line.amount_minor for line in lines)
    if actual != total_minor:
        raise SplitError(
            f"Die Summe der Splits ({actual}) entspricht nicht dem Betrag ({total_minor})."
        )


def total_of(lines: list[SplitLine]) -> int:
    return sum(line.amount_minor for line in lines)
